### PR TITLE
Improved frame decoder module

### PR DESCRIPTION
**Pull request by bsekachev**
### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have added a description of my changes into the [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [x] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

### CODE REVIEW

File: cvat-core/src/server-proxy.ts [hunk 1]
```diff
@@ -1958,7 +1955,7 @@ async function getOrganizations() {
     return response.results;
 }
 
-async function createOrganization(data) {
+async function createOrganization(data: SerializedOr) {
```
Comment (klakhov) [inline cvat-core/src/server-proxy.ts:1958]:
```suggestion
async function createOrganization(data: SerializedOrganization) {
```

File: cvat-core/src/frames.ts [hunk 2]
```diff
@@ -707,80 +444,93 @@ export async function getFrame(
...
 
         // limit of decoded frames cache by 2GB
-        const decodedBlocksCacheSize = Math.floor(2147483648 / (mean + stdDev) / 4 / chunkSize) || 1;
-
+        const decodedBlocksCacheSize = Math.min(
+            Math.floor((2048 * 1024 * 1024) / ((mean + stdDev) * 4 * chunkSize)) || 1, 10,
+        );
         frameDataCache[jobID] = {
-            meta,
+            meta: updatedMeta,
```
Comment (klakhov) [inline cvat-core/src/frames.ts:467]: There is some type error with updatedMeta

File: cvat-core/src/frames.ts [hunk 3]
```diff
@@ -3,28 +3,35 @@
...
-    activeChunkRequest: null;
-    nextChunkRequest: null;
+    activeChunkRequest: Promise<void> | null;
+    activeContextRequest: Promise<Record<number, ImageBitmap>> | null;
+    contextCache: Record<number, {
+        data: Record<number, ImageBitmap>;
+        timestamp: number;
+        size: number;
+    }>;
+    getChunk: (chunkNumber: number, quality: 'original' | 'compressed') => Promise<ArrayBuffer>;
```
Comment (klakhov) [inline cvat-core/src/frames.ts:34]: Maybe create enum for `quality`, Ive seen `quality: 'original' | 'compressed'` in several places

File: cvat-core/src/frames.ts [hunk 4]
```diff
@@ -157,233 +157,173 @@ export class FrameData {
...
+                return nextChunkNumber;
+            }
+
+            if (frame) {
+                if (decodeForward && decodedBlocksCacheSize > 1 && !frameDataCache[this.jobID].activeChunkRequest) {
+                    const nextChunkNumber = findTheNextNotDecodedChunk(this.number);
+                    const predecodeChunksMax = Math.floor(decodedBlocksCacheSize / 2);
+                    if (nextChunkNumber * chunkSize <= stopFrame &&
+                        nextChunkNumber <= chunkNumber + predecodeChunksMax) {
+                        provider.cleanup(1);
```
Comment (klakhov) [inline cvat-core/src/frames.ts:212]: Isn't `1` default argument for `cleanup`?

File: cvat-core/src/frames.ts [hunk 5]
```diff
@@ -707,80 +444,93 @@ export async function getFrame(
...
+            updatedMeta.frames.map((x) => (x.width * x.height - mean) ** 2).reduce((a, b) => a + b) /
+            updatedMeta.frames.length,
         );
 
         // limit of decoded frames cache by 2GB
-        const decodedBlocksCacheSize = Math.floor(2147483648 / (mean + stdDev) / 4 / chunkSize) || 1;
-
+        const decodedBlocksCacheSize = Math.min(
+            Math.floor((2048 * 1024 * 1024) / ((mean + stdDev) * 4 * chunkSize)) || 1, 10,
+        );
```
Comment (klakhov) [inline cvat-core/src/frames.ts:465]: Can you please explain these calculations?

File: .vscode/launch.json [hunk 6]
```diff
@@ -520,6 +520,7 @@
             "type": "node",
             "request": "launch",
             "name": "jest debug",
+            "cwd": "${workspaceFolder}/cvat-core",
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply .vscode/launch.json:523]: Yes, it did not work without it

File: cvat-core/src/frames.ts (hunk 5 above)
Comment (bsekachev) [reply cvat-core/src/frames.ts:465]: Average number of pixels (mean) * number of channels per image (4) * chunks size.
This is maximum number of decoded chunks we can keep in memory 2048 Mb

stdDev is 0 for images with the same size and > 0 for images with different size, also statistics characteristic

File: cvat-core/src/frames.ts (hunk 4 above)
Comment (bsekachev) [reply cvat-core/src/frames.ts:212]: Actually yes, but explicitly specifying it here is not extra

File: cvat-core/src/frames.ts [hunk 7]
```diff
@@ -157,233 +157,173 @@ export class FrameData {
...
     }
 
-    async data(onServerRequest = () => {}) {
+    async data(onServerRequest = () => {}): Promise<ImageBitmap | Blob> {
         const result = await PluginRegistry.apiWrapper.call(this, FrameData.prototype.data, onServerRequest);
         return result;
     }
 
     get imageData() {
+        // todo: check where it is used
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-core/src/frames.ts]: Done, used in opencv

File: cvat-core/src/frames.ts (hunk 2 above)
Comment (bsekachev) [reply cvat-core/src/frames.ts:457]: Fixed

File: cvat-core/src/frames.ts (hunk 3 above)
Comment (bsekachev) [reply cvat-core/src/frames.ts]: Added

**Review by klakhov — APPROVED** [review]

**Pull request merged**